### ISSUE TITLE
Deprecated usage of id in __hash__ methods

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1842]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#1842|
|Type|defect|
|Created|2006-06-22 10:28:20Z|

I tested Python2.5 beta1 and see that id() changed to return always a positive number (and possibly a long). In http://www.python.org/download/releases/2.5/NEWS.txt you can see they warn against the use of id in a !__hash!__ function, which needs an int.

I count 3 occurences of this use: in web.microdom, spread.flavors, and python.usage. The last one is very harmful because it prevents me from running trial :).

<details><summary>Searchable metadata</summary>

```
trac-id__1842 1842
type__defect defect
reporter__therve therve
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__hash_id hash id
time__1150972100000000 1150972100000000
changetime__1151728619000000 1151728619000000
version__None None
owner__ 
cc__therve cc__spiv
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

Without looking too closely, replacing 
```python
    def __hash__(self):
        return id(self)
```

with: 
```python
    def __hash__(self):
        return id(self) % sys.maxint
```

would probably be a simple fix (it seems happy enough with longs, so long as they fit in an int).

Perhaps Python 2.5 should do that automatically?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve commented|
|-|-|

Well it's not really a fix, it's a hack. I think we should find something better: for each object, we have attributes that can be used. For example for microdom.Node, someting like this should work:

```python
def __hash__(self):
    h = 0
    for child in self.childNodes:
        h ^= hash(child)
    return h
```

And for usage.Option:

```python
def __hash__(self):
    return hash(self.shortOpt)
```

For spread.RemoteCache, it's less easy to choose.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

(In [17388]) Fix hashability of three classes in Python 2.5

Refs [#1842](https://github.com/twisted/twisted/issues/1842)
Refs [#1867](https://github.com/twisted/twisted/issues/1867)

Options and RemoteCache basically just want to be hashable by identity: give
them an implementation that is almost the same as the previous one, but
which truncates to "short" int so that Python does not get upset.  We could
probably do something which exploited the negative "address space" (hashes
are signed) but it seems extremely unlikely that any difference due to this
would be noticable in practice.

microdom's Node class has its hash method removed entirely since it was
completely superfluous: the Node class defines no custom comparison methods
and so instances of Node will only compare equal to themselves, and the
default hash is by identity anyway, which is what the implementation being
removed was duplicated.

**Comment by twisted-trac** [comment]
|<img alt="foom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|foom set status to `closed`|
|-|-|

(In [17448]) Fix almost all python 2.5 issues.

Merging branches/python-two-point-five-1867.

Authors: mostly exarkun, some jknight
Reviewer: jml
Closes [#1867](https://github.com/twisted/twisted/issues/1867)

In detail:

Fix hashability of three classes in Python 2.5: id() now returns only
positive numbers, but no longer always short ints, so returning
id(self) as __hash__ breaks. Closes [#1842](https://github.com/twisted/twisted/issues/1842)

Truncate mtime and atime floats to be integers in conch's unix sftp
server implementation. (They changed to be floats by default in
python). The interface is still underspecified. Refs [#1860](https://github.com/twisted/twisted/issues/1860).

Fix Newpb failure handling for new-style exceptions. Use t.p.r.qual to
determine exception type names. Closes [#1861](https://github.com/twisted/twisted/issues/1861)

Suppress deprecation warning for raising a string exception in
test_failure.FailureTestCase.testStringExceptions. Closes [#1862](https://github.com/twisted/twisted/issues/1862)

Refactor testPBFailures into multiple test methods to make it easier
to debug. Closes [#1863](https://github.com/twisted/twisted/issues/1863)

Pass absolute path to interpreter as argv[0] in stdio tests. Using an
absolute path lets Python's import machinery figure out what's going
on. Closes [#1864](https://github.com/twisted/twisted/issues/1864)

Add new-style class support to jelly (absolutely necessary because
exceptions are now newstyle classes). Closes [#1865](https://github.com/twisted/twisted/issues/1865)

Add new-style exception support to Failure. 
Closes [#1858](https://github.com/twisted/twisted/issues/1858). Closes [#1859](https://github.com/twisted/twisted/issues/1859).

Fix twisted.python.zipstream to account for the removal of file_offset
from the ZipInfo class. Closes [#1866](https://github.com/twisted/twisted/issues/1866).

Use reflect.qual on warning class when logging warnings.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
